pkg/resourcequota/resource_quota_controller.go

```diff
@@ -0,0 +1,194 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourcequota
+
+import (
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api/resource"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/golang/glog"
+)
+
+// ResourceQuotaManager is responsible for tracking quota usage status in the system
+type ResourceQuotaManager struct {
+	kubeClient client.Interface
+	syncTime   <-chan time.Time
+
+	// To allow injection of syncUsage for testing.
+	syncHandler func(quota api.ResourceQuota) error
+}
+
+// NewResourceQuotaManager creates a new ResourceQuotaManager
+func NewResourceQuotaManager(kubeClient client.Interface) *ResourceQuotaManager {
+
+	rm := &ResourceQuotaManager{
+		kubeClient: kubeClient,
+	}
+
+	// set the synchronization handler
+	rm.syncHandler = rm.syncResourceQuota
+	return rm
+}
+
+// Run begins watching and syncing.
+func (rm *ResourceQuotaManager) Run(period time.Duration) {
+	rm.syncTime = time.Tick(period)
+	go util.Forever(func() { rm.synchronize() }, period)
+}
+
+func (rm *ResourceQuotaManager) synchronize() {
+	var resourceQuotas []api.ResourceQuota
+	list, err := rm.kubeClient.ResourceQuotas(api.NamespaceAll).List(labels.Everything())
+	if err != nil {
+		glog.Errorf("Synchronization error: %v (%#v)", err, err)
+	}
+	resourceQuotas = list.Items
+	wg := sync.WaitGroup{}
+	wg.Add(len(resourceQuotas))
+	for ix := range resourceQuotas {
```

Comment:
http://www.goinggo.net/2014/01/concurrency-goroutines-and-gomaxprocs.html?m=1

Great article.  Short answer is that this is fine to do because go routines are different than threads.  This would have been a bad thing if we modified runtime.GOMAXPROCS

Sent from my iPhone

> On Jan 27, 2015, at 4:02 AM, Eric Tune <notifications@github.com> wrote:
> 
> In pkg/resourcequota/resource_quota_controller.go:
> 
> > +// Run begins watching and syncing.
> > +func (rm *ResourceQuotaManager) Run(period time.Duration) {
> > +	rm.syncTime = time.Tick(period)
> > +	go util.Forever(func() { rm.synchronize() }, period)
> > +}
> > +
> > +func (rm *ResourceQuotaManager) synchronize() {
> > +	var resourceQuotas []api.ResourceQuota
> > +	list, err := rm.kubeClient.ResourceQuotas(api.NamespaceAll).List(labels.Everything())
> > +	if err != nil {
> > +		glog.Errorf("Synchronization error: %v (%#v)", err, err)
> > +	}
> > +	resourceQuotas = list.Items
> > +	wg := sync.WaitGroup{}
> > +	wg.Add(len(resourceQuotas))
> > +	for ix := range resourceQuotas {
> If there are 1000 namespaces times 5 ResourceQuotas per namespace , is it a good idea to launch 5000 goroutines? Honest question. I probably wouldn't do that if it was threads, but maybe goroutines are different.
> 
> —
> Reply to this email directly or view it on GitHub.
>